### ISSUE TITLE
Modified jvm.options prevent node from starting when upgrading from 7.5.1 to 7.7.0

### ISSUE THREAD

**Issue by ddolcimascolo** [elastic/elasticsearch#57356]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.7.0

**Plugins installed**: []

**JVM version** (`java -version`): The bundled JDK 14 (version below)

```
$ /usr/share/elasticsearch/jdk/bin/java -version
openjdk version "14" 2020-03-17
OpenJDK Runtime Environment AdoptOpenJDK (build 14+36)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14+36, mixed mode, sharing)
```

**OS version** (`uname -a` if on a Unix-like system): Ubuntu Server 18.04

```
$ uname -a
Linux 4.15.0-99-generic #100-Ubuntu SMP Wed Apr 22 20:32:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.4 LTS"
```

**Description of the problem including expected versus actual behavior**:

_Preamble_:
In versions prior to 7.7.0 we used to edit the file `/etc/elasticsearch/jvm.options` directly to change the heap size, as documented. This file is marked as a configuration file by the Debian package and as such, triggers a specific workflow in `dpkg` when the package is upgraded (depending on various advanced options that you might pass to `apt`) which by default asks you whether you want to keep your modified file, or upgrade to the new one. The problem gets worse when you upgrade the package in an unattended way, typically via a configuration manager (we use Ansible), because `dpkg` will keep the old version without asking.

_The problem_:
As a matter of fact, upgrading Elasticsearch in these conditions silently ignore any upgrades to the `jvm.options` file, which was sad but worked fine until 7.7.0.
This version bundles a JDK 14 in which the CMS GC does not exist anymore and (on my system at least) the JVM defaults to the serial (single-threaded) GC (proof below)

```
$ /usr/share/elasticsearch/jdk/bin/java -XX:+PrintCommandLineFlags -version
-XX:InitialHeapSize=64637824 -XX:MaxHeapSize=1034205184 -XX:MinHeapSize=6815736 -XX:+PrintCommandLineFlags -XX:ReservedCodeCacheSize=251658240 -XX:+SegmentedCodeCache -XX:+UseCompressedClassPointers -XX:+UseCompressedOops -XX:+UseSerialGC 
openjdk version "14" 2020-03-17
OpenJDK Runtime Environment AdoptOpenJDK (build 14+36)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14+36, mixed mode, sharing)
```
The previous flavor of the `jvm.options` file enabled the CMS GC, but it does not exist anymore so the JVM uses the default one. This prevents the node from starting after the upgrade because of a failing bootstrap check.

**Steps to reproduce**:

 1. Install Elasticsearch version 7.5.1 (I suppose any version < 7.7.0 would reproduce) using the official Debian packages on Ubuntu 18.04
 2. Edit the `/etc/elasticsearch/jvm.options` file to increase the heap size
 2. Verify that the node starts
 3. Upgrade to Elasticsearch 7.7.0 using APT, keep your old version of the `jvm.options` file (answer "N" when asked)
 4. Start the node, see it fail in the logs

**Additional information**

The new system for setting JVM options with files in the `/etc/elasticsearch/jvm.options.d` directory is the correct solution to this. I've now migrated the nodes to use a file in this folder.
However, the problem I describe here is still an issue and make upgrading painful, at least you should be able to show a big warning when the package is upgraded that the user MUST upgrade the `jvm.options` file because of the upgraded bundled JDK. Better yet, detect this problem and guide the user towards resolving this "the proper way". I personally spent a few hours digging into the changelog in a hope to find what the problem is...

**Provide logs (if relevant)**:

```
[2020-05-29T09:06:50,490][INFO ][o.e.e.NodeEnvironment    ] [es-002.stordata.dev] using [1] data paths, mounts [[/ (/dev/sda2)]], net usable_space [38.6gb], net total_space [62.7gb], types [ext4]
[2020-05-29T09:06:50,499][INFO ][o.e.e.NodeEnvironment    ] [es-002.stordata.dev] heap size [1.8gb], compressed ordinary object pointers [true]
[2020-05-29T09:06:51,312][INFO ][o.e.n.Node               ] [es-002.stordata.dev] node name [es-002.stordata.dev], node ID [nfaUfK3uST-UxMCKQBDYAQ], cluster name [elk-clus-00]
[2020-05-29T09:06:51,318][INFO ][o.e.n.Node               ] [es-002.stordata.dev] version[7.7.0], pid[626], build[default/deb/81a1e9eda8e6183f5237786246f6dced26a10eaf/2020-05-12T02:01:37.602180Z], OS[Linux/4.15
.0-99-generic/amd64], JVM[AdoptOpenJDK/OpenJDK 64-Bit Server VM/14/14+36]
[2020-05-29T09:06:51,318][INFO ][o.e.n.Node               ] [es-002.stordata.dev] JVM home [/usr/share/elasticsearch/jdk]
[2020-05-29T09:06:51,319][INFO ][o.e.n.Node               ] [es-002.stordata.dev] JVM arguments [-Xshare:auto, -Des.networkaddress.cache.ttl=60, -Des.networkaddress.cache.negative.ttl=10, -XX:+AlwaysPreTouch, -
Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -XX:+ShowCodeDetailsInExceptionMessages, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true,
 -Dio.netty.recycler.maxCapacityPerThread=0, -Dio.netty.allocator.numDirectArenas=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -Djava.locale.providers=SPI,COMPAT, -Xms1972m, -Xmx1972m, -XX:+
UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -Djava.io.tmpdir=/tmp/elasticsearch-4206078949703173838, -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=/var/lib
/elasticsearch, -XX:ErrorFile=/var/log/elasticsearch/hs_err_pid%p.log, -Xlog:gc*,gc+age=trace,safepoint:file=/var/log/elasticsearch/gc.log:utctime,pid,tags:filecount=32,filesize=64m, -Xms1972m, -Xmx1972m, -XX:M
axDirectMemorySize=1033895936, -Des.path.home=/usr/share/elasticsearch, -Des.path.conf=/etc/elasticsearch, -Des.distribution.flavor=default, -Des.distribution.type=deb, -Des.bundled_jdk=true]
[2020-05-29T09:06:57,156][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [aggs-matrix-stats]
[2020-05-29T09:06:57,159][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [analysis-common]
[2020-05-29T09:06:57,162][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [constant-keyword]
[2020-05-29T09:06:57,162][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [flattened]
[2020-05-29T09:06:57,162][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [frozen-indices]
[2020-05-29T09:06:57,163][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [ingest-common]
[2020-05-29T09:06:57,163][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [ingest-geoip]
[2020-05-29T09:06:57,163][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [ingest-user-agent]
[2020-05-29T09:06:57,164][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [lang-expression]
[2020-05-29T09:06:57,165][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [lang-mustache]
[2020-05-29T09:06:57,166][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [lang-painless]
[2020-05-29T09:06:57,166][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [mapper-extras]
[2020-05-29T09:06:57,166][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [parent-join]
[2020-05-29T09:06:57,166][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [percolator]
[2020-05-29T09:06:57,167][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [rank-eval]
[2020-05-29T09:06:57,167][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [reindex]
[2020-05-29T09:06:57,167][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [repository-url]
[2020-05-29T09:06:57,167][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [search-business-rules]
[2020-05-29T09:06:57,173][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [spatial]
[2020-05-29T09:06:57,174][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [systemd]
[2020-05-29T09:06:57,174][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [tasks]
[2020-05-29T09:06:57,174][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [transform]
[2020-05-29T09:06:57,174][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [transport-netty4]
[2020-05-29T09:06:57,175][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [vectors]
[2020-05-29T09:06:57,176][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-analytics]
[2020-05-29T09:06:57,176][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-async-search]
[2020-05-29T09:06:57,177][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-autoscaling]
[2020-05-29T09:06:57,177][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-ccr]
[2020-05-29T09:06:57,178][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-core]
[2020-05-29T09:06:57,178][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-deprecation]
[2020-05-29T09:06:57,178][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-enrich]
[2020-05-29T09:06:57,181][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-eql]
[2020-05-29T09:06:57,182][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-graph]
[2020-05-29T09:06:57,182][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-identity-provider]
[2020-05-29T09:06:57,182][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-ilm]
[2020-05-29T09:06:57,183][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-logstash]
[2020-05-29T09:06:57,185][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-ml]
[2020-05-29T09:06:57,186][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-monitoring]
[2020-05-29T09:06:57,186][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-ql]
[2020-05-29T09:06:57,187][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-rollup]
[2020-05-29T09:06:57,187][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-security]
[2020-05-29T09:06:57,188][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-sql]
[2020-05-29T09:06:57,188][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-voting-only-node]
[2020-05-29T09:06:57,194][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] loaded module [x-pack-watcher]
[2020-05-29T09:06:57,195][INFO ][o.e.p.PluginsService     ] [es-002.stordata.dev] no plugins loaded
[2020-05-29T09:07:09,620][INFO ][o.e.x.s.a.s.FileRolesStore] [es-002.stordata.dev] parsed [0] roles from file [/etc/elasticsearch/roles.yml]
[2020-05-29T09:07:11,311][INFO ][o.e.x.m.p.l.CppLogMessageHandler] [es-002.stordata.dev] [controller/795] [Main.cc@110] controller (64 bit): Version 7.7.0 (Build a8939d3da43f33) Copyright (c) 2020 Elasticsearch BV
[2020-05-29T09:07:13,254][INFO ][o.e.d.DiscoveryModule    ] [es-002.stordata.dev] using discovery type [zen] and seed hosts providers [settings]
[2020-05-29T09:07:15,639][INFO ][o.e.n.Node               ] [es-002.stordata.dev] initialized
[2020-05-29T09:07:15,642][INFO ][o.e.n.Node               ] [es-002.stordata.dev] starting ...
[2020-05-29T09:07:15,948][INFO ][o.e.t.TransportService   ] [es-002.stordata.dev] publish_address {10.78.7.228:9300}, bound_addresses {10.78.7.228:9300}
[2020-05-29T09:07:22,938][INFO ][o.e.b.BootstrapChecks    ] [es-002.stordata.dev] bound or publishing to a non-loopback address, enforcing bootstrap checks
[2020-05-29T09:07:22,990][ERROR][o.e.b.Bootstrap          ] [es-002.stordata.dev] node validation exception
[1] bootstrap checks failed
[1]: JVM is using the serial collector but should not be for the best performance; either it's the default for the VM [OpenJDK 64-Bit Server VM] or -XX:+UseSerialGC was explicitly specified
[2020-05-29T09:07:23,005][INFO ][o.e.n.Node               ] [es-002.stordata.dev] stopping ...
[2020-05-29T09:07:23,029][INFO ][o.e.n.Node               ] [es-002.stordata.dev] stopped
[2020-05-29T09:07:23,030][INFO ][o.e.n.Node               ] [es-002.stordata.dev] closing ...
[2020-05-29T09:07:23,082][INFO ][o.e.n.Node               ] [es-002.stordata.dev] closed
[2020-05-29T09:07:23,084][INFO ][o.e.x.m.p.NativeController] [es-002.stordata.dev] Native controller process has stopped - no new native processes can be started
```

Thanks a lot !

David

**Comment by ddolcimascolo** [comment]
Hey guys, am I really the only one hitting this? I'll be glad to hear back from you to understand what we did wrong...

**Comment by rjernst** [comment]
@ddolcimascolo As you noted, `jvm.options.d` is the correct long term solution for this problem, which we have already implemented. It looks like you are hitting an edge case since the jvm will default to serial GC in very few cases (are you on a single core?). And there isn't really anything we can do as far as upgrade warnings, especially in non-interactive scenarios. Configuration file conflicts are the realm of package managers. The change to use `jvm.options.d` makes those conflicts easier to deal with in the future, but the existing single `jvm.options` file conflicting across upgrades was a known issue, and exactly why we added the directory based approach.

We are sorry for your troubles, and happy that it looks like you figured out the correct workaround. Given that,  and our desire to leave configuration conflicts to package managers, I hope you don't mind that I close this issue.
